Improved binning for time_bins

KEichenseer opened the pull request:
# time_bins

## Description

- Updating the binning in the `time_bins` functionto use a dynamic programming approach which generates a cost matrix for potential partitionings (cost being (target_bin_size - actual_bin_size)^2) and selects the lowest-cost partitionings. 
- Removes edge effects sometimes present in the previous approach and generally brings average bin sizes closer to the user-specified `size`. 
- Sometimes this better approximation of desired average bin length comes at the cost of higher sd, but I think that's OK - the user can specify different bin sizes if the current size forces relatively uneven spacings. 
- Adding a `n_bins` option where the user can specify a desired number of bins instead of bin `size`.

## Motivation and Context
Improve binning for `time_bins()` following discussion in #18 .

## Checklist before requesting a review
- [x] I have read palaeoverse's [standards and structure](https://palaeoverse.org/articles/structure-and-standards.html) 
- [x] I have performed a self-review of my code.
- [x] I have added function documentation.
- [x] I have provided sufficient examples of implementation.
- [x] If it is a core feature, I have added thorough tests.

willgearty commented on R/time_bins.R:
```diff
@@ -293,74 +304,124 @@ time_bins <- function(interval = "Phanerozoic", rank = "stage", size = NULL,
+    } else {
+      # Throw error if n_bins is too large
+      if (n_bins > nrow(df)) {
+        stop("`n_bins` can't be greater than the number intervals.")
```
```suggestion
        stop("`n_bins` can't be greater than the number of intervals.")
```

KEichenseer replied:
Implemented

willgearty commented on R/time_bins.R:
```diff
@@ -293,74 +304,124 @@ time_bins <- function(interval = "Phanerozoic", rank = "stage", size = NULL,
+                            round(size, digits = 2),  "Myr.\n")
+    } else {
+      message_head <- paste0("Number of equal length time bins was set to ",
+                            n_bins, ".\n")
```
```suggestion
                             n_bins, ".\n")
```

KEichenseer replied:
Implemented

willgearty commented on R/time_bins.R:
```diff
@@ -293,74 +304,124 @@ time_bins <- function(interval = "Phanerozoic", rank = "stage", size = NULL,
-        intervals[i] <- toString(df[lower[i]:upper[i], c("interval_name")])
+    }
+    if (n_bins > 1) {
+    # updated bin size based on determined number of bins
```
Everything in this `if` should be indented once

KEichenseer replied:
Implemented

willgearty commented on R/time_bins.R:
```diff
@@ -377,12 +438,12 @@ time_bins <- function(interval = "Phanerozoic", rank = "stage", size = NULL,
     }
-    if (is.numeric(size)) {
+    if (is.numeric(size) || is.numeric(n_bins)) {
       title(paste(
```
I think we should use paste0 here, otherwise we get a space before the last parenthesis

KEichenseer replied:
Implemented

willgearty commented on R/time_bins.R:
```diff
@@ -293,74 +304,124 @@ time_bins <- function(interval = "Phanerozoic", rank = "stage", size = NULL,
+                            n_bins, ".\n")
+    }
+    message(
+      paste(message_head,
```
Should change to using paste0 here, otherwise we get a space before the "time bins were generated" line

KEichenseer replied:
Implemented

willgearty reviewed (approved):
This looks really good to me @KEichenseer, thanks for figuring this out and implementing it! Just some very minor style and print comments.

KEichenseer commented:
> This looks really good to me @KEichenseer, thanks for figuring this out and implementing it! Just some very minor style and print comments.

Thanks @willgearty! I've implemented your changes and some additional lint fixes, hope all is clean now.

LewisAJones commented on tests/testthat/test-time_bins.R:
```diff
@@ -4,8 +4,10 @@ test_that("time_bins() works with macrostrat", {
-  expect_equal(nrow(time_bins(size = 10,
+  expect_message(
+    expect_equal(nrow(time_bins(size = 10,
                               scale = "North american land mammal ages")), 7)
```
Check indentation throughout test file

LewisAJones reviewed (changes requested):
@KEichenseer, this looks great to me. Really impressive that you found a solution to this which is also fast. One of my fears with trying to find a solution to this was that it would end up taking a while to run, but this is very efficient!

I've got a few minor comments, mainly relating to the documentation and code formatting (don't forget to update `NEWs.md` too). I really appreciate that you went through the effort of substantially commenting your code - it made it much easier to follow.

Exciting to finally have this in the package - don't forget to add yourself to the Developer(s) section!

Unrelated to this PR, but it got me thinking about what our long-term plan should be with the internal GTS data tables. Should we look to include another version in the future, keep it as it is, or eventually remove the GTS as they become increasingly outdated? Something for us to think about.

KEichenseer commented on R/time_bins.R:
```diff
@@ -7,9 +7,11 @@
-#' 10 Myr. However, users may also want to consider grouping stages based on
-#' other reasoning e.g. availability of outcrop (see Dean et al. 2020).
+#' function will generate bins that have a mean bin length close to
+#' 10 Myr. Similarly, for a given bin number n, the function will generate n
```
I slightly changed that to work with `n_bins` rather than `n`. Also, should we use `n_bins` rather than *n_bins*?

LewisAJones replied:
Happy with that, `n_bins` is fine! This was also fine tbh, I was just being picky because you didn't leave me much to comment on 😄

KEichenseer replied:
Ok great, will push as soon as the checks pass and you can see if you find more ;-)
Indentation is my nemesis :-D

KEichenseer commented on R/time_bins.R:
```diff
@@ -135,9 +146,9 @@ time_bins <- function(interval = "Phanerozoic", rank = "stage", size = NULL,
 
   if (is.data.frame(scale) &&
-      any(!c("interval_name", "max_ma", "min_ma") %in% colnames(scale))) {
+        any(!c("interval_name", "max_ma", "min_ma") %in% colnames(scale))) {
```
Thanks for finding that, have tried to correct those indentation issues with cut - paste now.

LewisAJones commented on NEWS.md:
```diff
@@ -8,6 +8,7 @@
 * Added lat_bins_area (#114)
 * Replaced lat_bins with lat_bins_degrees
 * Made all functions tibble-friendly (#118)
+* Made bin durations more even and added n_bins option in time_bins (#131)
```
palaeoverse 1.4.0 has already been released, this should be under a new header of `# Development version`

LewisAJones commented:
> Thanks @LewisAJones, I've implemented your changes, added myself as co-author (do you want to add yourself as reviewer?) and a sentence to the NEWS.md.

Thanks just one small comment related to the `NEWS.md`! I'm fine with not being added as a reviewer. I guess as a developer, the assumption is that you've reviewed the code too.

> I don't think we should remove versions of GTS as this might break scripts of users when they update palaeoverse. Adding a new one every couple of years would certainly be welcome. Changing the default to the newest one should be fine as long as the older versions stay available.

You can currently always get the most recent version of the GTS via Macrostrat, e.g. 

```
palaeoverse::time_bins(scale = "International Ages")
```

I guess we could use this to update/add new static versions of the GTS in the future - but I guess we don't want to end up with having too many different versions in the package. Perhaps we should only include new versions for instances when there are major changes, such as changes to names and the creation of additional stages (such as the Holocene being split a few years ago).

KEichenseer commented:
> > Thanks @LewisAJones, I've implemented your changes, added myself as co-author (do you want to add yourself as reviewer?) and a sentence to the NEWS.md.
> 
> Thanks just one small comment related to the `NEWS.md`! I'm fine with not being added as a reviewer. I guess as a developer, the assumption is that you've reviewed the code too.
> 
> > I don't think we should remove versions of GTS as this might break scripts of users when they update palaeoverse. Adding a new one every couple of years would certainly be welcome. Changing the default to the newest one should be fine as long as the older versions stay available.
> 
> You can currently always get the most recent version of the GTS via Macrostrat, e.g.
> 
> ```
> palaeoverse::time_bins(scale = "International Ages")
> ```

I see, perhaps it should be spelled out in the documentation that Macrostrat has options for the most recent GTS?

> 
> I guess we could use this to update/add new static versions of the GTS in the future - but I guess we don't want to end up with having too many different versions in the package. Perhaps we should only include new versions for instances when there are major changes, such as changes to names and the creation of additional stages (such as the Holocene being split a few years ago).

Yes, that works.

LewisAJones commented:
> I see, perhaps it should be spelled out in the documentation that Macrostrat has options for the most recent GTS?

I've added some documentation [https://github.com/palaeoverse/palaeoverse/pull/131/commits/4e739e3b70c5375b1e48a2f1ebcc8058cdd78b9e](https://github.com/palaeoverse/palaeoverse/pull/131/commits/4e739e3b70c5375b1e48a2f1ebcc8058cdd78b9e)... 

If you're happy with this, I think we can go ahead and merge! Thanks for all the great work @KEichenseer.

KEichenseer commented:
> > I see, perhaps it should be spelled out in the documentation that Macrostrat has options for the most recent GTS?
> 
> I've added some documentation [4e739e3](https://github.com/palaeoverse/palaeoverse/commit/4e739e3b70c5375b1e48a2f1ebcc8058cdd78b9e)...
> 
> If you're happy with this, I think we can go ahead and merge! Thanks for all the great work @KEichenseer.

Thanks @LewisAJones, yes I'm happy with this, please go ahead and merge.
